### ISSUE TITLE
fallback: :exception with JSON returns empty error message

### ISSUE THREAD

**Issue by iampeter** [gonzalo-bulnes/simple_token_authentication#201]
I am forcing JSON response and using `fallback: :exception`, and the response is 

```
{
"error": ""
}
```

although the code in `ExceptionCallbackHandler` says
```
throw(:warden, scope: entity.name_underscore.to_sym) if controller.send("current_#{entity.name_underscore}").nil?
```

Why doesn't it return any message?

I have tried doing it this way:
```
  class UnauthenticatedException < Exception
  end

  rescue_from UnauthenticatedException do |exception|
    render json: { error: "Authentication error", code: 1 }, status: 401
  end

  class ApiExceptionFallbackHandler
    # Notifies the failure of authentication to Warden in the same DEvise does.
    # Does result in an HTTP 401 response in a Devise context.
    def fallback!(controller, entity)
      raise UnauthenticatedException.new if controller.send("current_#{entity.name_underscore}").nil?
    end
  end

  @@fallback_authentication_handler = ApiExceptionFallbackHandler.new

  acts_as_token_authentication_handler_for ApiKey
```

and it works, but am bot sure if raising and Exception instead of throw(:warden) is a legitimate way here.

**Comment by gonzalo-bulnes** [comment]
Hi @iampeter,

I'm reading about this, but it seems (see [these][example-1] [examples][example-2]) that you can provide a message when `thow`ing the `:warden` symbol. I would try:

```ruby
# lib/simple_token_authentication/exception_fallback_handler.rb

module SimpleTokenAuthentication
  class ExceptionFallbackHandler
    # Notifies the failure of authentication to Warden in the same Devise does.
    # Does result in an HTTP 401 response in a Devise context.
    def fallback!(controller, entity)
      throw(:warden, scope: entity.name_underscore.to_sym, message: 'Invalid authentication credentials.') if controller.send("current_#{entity.name_underscore}").nil?
    end
  end
end
```

Indeed, `message: :invalid_token_authentication_credentials` might be enough if you define the message in your locales file ([idea][locales]).

Unless I'm wrong, Devise [will get this message all the way through your JSON response][message_handling].

  [locales]: https://github.com/plataformatec/devise/commit/dff7891b97d785d228a15dff5c1acdb0da34605f#diff-37c9b835927e3c363588f172de9bb6faL14
  [message_handling]: https://github.com/plataformatec/devise/blob/v3.5.2/lib/devise/failure_app.rb#L76
  [example-1]: https://github.com/plataformatec/devise/blob/v3.5.2/lib/devise/hooks/timeoutable.rb#L24
  [example-2]: https://github.com/plataformatec/devise/blob/v3.5.2/lib/devise/hooks/activatable.rb#L8

**Comment by gonzalo-bulnes** [comment]
About raising an exception vs. throwing a symbol: I found all _three_ answers to [this SO question][so] helpful.

In this case, having to handle uncorrect or missing credentials is _expected_, and undoubtly part of the token authentication job. That's why I would avoid relying on exceptions to do it.
That being said, I was not familiar with the distinction myself and I think I'll refactor some code [here][1] and [there][2] ; )

  [so]: https://stackoverflow.com/questions/51021/what-is-the-difference-between-raising-exceptions-vs-throwing-exceptions-in-ruby
  [1]: https://github.com/gonzalo-bulnes/dredd-rack/blob/v0.7.0/lib/dredd/rack/runner.rb#L95
  [2]: https://github.com/gonzalo-bulnes/dredd-rack/blob/v0.7.0/lib/dredd/rack/runner.rb#L85

**Comment by gonzalo-bulnes** [comment]
Hi @iampeter,

Were you able to try passing a message when throwing `:warden`?

**Comment by iampeter** [comment]
@gonzalo-bulnes nope. that's why I'm doing it like mentioned in my original comment

**Comment by gkunwar** [comment]
@iampeter @gonzalo-bulnes Is this issue is fixed and released?

**Comment by gonzalo-bulnes** [comment]
Hi @gkunwar,

No, I haven't been able to dig further and understand how Devise does handle the error messages through Warden. If you can progress towards understanding what is missing in the suggestion I wrote above, I'd be glad to work with you to make that a releasable fix. (@krazedkrish That's true for you too!)

While I find @iampeter's workaround fine and clean, it is a workaround and does not really fix the issue.

Indeed, another workaround could be to use an `after_failed_token_authentication` hook. That wouldn't fix the issue either, but is a more general piece of code that it could make sense to release independently. See #217.

In both cases, I'm happy to spend time understanding the issue and the Devise code better, but I need help and feedback from people experiencing the issue to make sure we don't release anything that wouldn't make sense in practice. Does it sound fair?
